### ISSUE TITLE
slice(<grouped sf>)

### ISSUE THREAD

**Issue by romainfrancois** [tidyverse/dplyr#6084]
I believe the problem we see with the rev dep of the `motif` package is about `slice()` not restoring grouped sf data frames. 

(this could probably be made simpler): 

``` r
library(stars)
#> Loading required package: abind
#> Loading required package: sf
#> Linking to GEOS 3.9.0, GDAL 3.2.1, PROJ 7.2.1
#> Registered S3 methods overwritten by 'stars':
#>   method             from
#>   st_bbox.SpatRaster sf  
#>   st_crs.SpatRaster  sf
library(sf)
library(motif)

landform = read_stars(system.file("raster/landforms.tif", package = "motif"), proxy = FALSE)
landform_cove = lsp_signature(landform,
                              type = "cove",
                              window = 200,
                              threshold = 0.5,
                              normalization = "pdf")

landform_dist = lsp_to_dist(landform_cove,
                            dist_fun = "jensen-shannon")
#> Metric: 'jensen-shannon' using unit: 'log2'; comparing: 9 vectors.

landform_hclust = hclust(landform_dist, method = "ward.D2")

clusters = cutree(landform_hclust, k = 6)
landform_grid_sf = lsp_add_clusters(landform_cove, clusters)

(grouped_sf <- landform_grid_sf %>%
  dplyr::filter(na_prop < 0.01) %>%
  dplyr::group_by(clust))
#> Simple feature collection with 6 features and 4 fields
#> Geometry type: POLYGON
#> Dimension:     XY
#> Bounding box:  xmin: -400176.1 ymin: -519756.5 xmax: -220176.1 ymax: -399756.5
#> Projected CRS: unnamed
#> # A tibble: 6 × 5
#> # Groups:   clust [4]
#>      id  na_prop signature       clust                                  geometry
#>   <int>    <dbl> <list>          <int>                             <POLYGON [m]>
#> 1     1 0.000275 <dbl [1 × 144]>     1 ((-400176.1 -399756.5, -340176.1 -399756…
#> 2     2 0.000325 <dbl [1 × 144]>     1 ((-340176.1 -399756.5, -280176.1 -399756…
#> 3     3 0.00118  <dbl [1 × 144]>     1 ((-280176.1 -399756.5, -220176.1 -399756…
#> 4     5 0.00025  <dbl [1 × 144]>     2 ((-400176.1 -459756.5, -340176.1 -459756…
#> 5     6 0.0013   <dbl [1 × 144]>     3 ((-340176.1 -459756.5, -280176.1 -459756…
#> 6     7 0.000075 <dbl [1 × 144]>     4 ((-280176.1 -459756.5, -220176.1 -459756…
class(grouped_sf)
#> [1] "sf"         "grouped_df" "tbl_df"     "tbl"        "data.frame"

grouped_sf %>%
  dplyr::slice_sample(n = 4, replace = TRUE)
#> # A tibble: 6 × 5
#> # Groups:   clust [4]
#>      id  na_prop signature       clust                                  geometry
#>   <int>    <dbl> <list>          <int>                             <POLYGON [m]>
#> 1     2 0.000325 <dbl [1 × 144]>     1 ((-340176.1 -399756.5, -280176.1 -399756…
#> 2     1 0.000275 <dbl [1 × 144]>     1 ((-400176.1 -399756.5, -340176.1 -399756…
#> 3     3 0.00118  <dbl [1 × 144]>     1 ((-280176.1 -399756.5, -220176.1 -399756…
#> 4     5 0.00025  <dbl [1 × 144]>     2 ((-400176.1 -459756.5, -340176.1 -459756…
#> 5     6 0.0013   <dbl [1 × 144]>     3 ((-340176.1 -459756.5, -280176.1 -459756…
#> 6     7 0.000075 <dbl [1 × 144]>     4 ((-280176.1 -459756.5, -220176.1 -459756…
```

<sup>Created on 2021-11-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by romainfrancois** [comment]
The rev dep for `motif` gives: 

````
> checking tests ... ERROR
  See below...

> checking installed package size ... NOTE
    installed size is  9.6Mb
    sub-directories of 1Mb or more:
      libs     6.3Mb
      raster   2.2Mb

── Test failures ─────────────────────────────────────────────────────────────────────────────── testthat ────

> library(testthat)
> library(stars)
Loading required package: abind
Loading required package: sf
Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 6.3.1
> library(sf)
> library(motif)
> 
> # prepare regular stars ---------------------------------------------------
> # landcover = read_stars(system.file("raster/landcover2015.tif", package = "motif"), proxy = FALSE)
> # landform = read_stars(system.file("raster/landform.tif", package = "motif"), proxy = FALSE)
> # landcoverold = read_stars(system.file("raster/landcover2001.tif", package = "motif"), proxy = FALSE)
> 
> # prepare stars proxy -----------------------------------------------------
> # landform_p = read_stars(system.file("raster/landform.tif", package = "motif"), proxy = TRUE)
> # landcover_p = read_stars(system.file("raster/landcover2015.tif", package = "motif"), proxy = TRUE)
> # landcoverold_p = read_stars(system.file("raster/landcover2001.tif", package = "motif"), proxy = TRUE)
> 
> # prep test extent --------------------------------------------------------
> # t_ext = st_bbox(c(xmin = -400000, xmax = -200000,
> #                 ymin = -600000, ymax = -400000),
> #               crs = st_crs(landcover))
> 
> # prepare regular stars ---------------------------------------------------
> # landcover = landcover[t_ext]
> # landform = landform[t_ext]
> # landcoverold = landcoverold[t_ext]
> landcover = read_stars(system.file("raster/landcover2015s.tif", package = "motif"), proxy = FALSE)
> landform = read_stars(system.file("raster/landforms.tif", package = "motif"), proxy = FALSE)
> landcoverold = read_stars(system.file("raster/landcover2001s.tif", package = "motif"), proxy = FALSE)
> 
> # prepare stars proxy -----------------------------------------------------
> # landform_p = st_crop(landform_p, t_ext)
> # landcover_p = st_crop(landcover_p, t_ext)
> # landcoverold_p = st_crop(landcoverold_p, t_ext)
> landcover_p = read_stars(system.file("raster/landcover2015s.tif", package = "motif"), proxy = TRUE)
> landform_p = read_stars(system.file("raster/landforms.tif", package = "motif"), proxy = TRUE)
> landcoverold_p = read_stars(system.file("raster/landcover2001s.tif", package = "motif"), proxy = TRUE)
> 
> # prepare poly ------------------------------------------------------------
> # ecoregions = read_sf(system.file("vector/ecoregions.gpkg", package = "motif"))
> # # st_crs(ecoregions) = st_crs(landform_p)
> # ecoregions = st_transform(ecoregions, st_crs(landform_p))
> # ecoregions = st_crop(ecoregions, t_ext)
> # ecoregions = st_cast(ecoregions, "MULTIPOLYGON")
> ecoregions = read_sf(system.file("vector/ecoregionss.gpkg", package = "motif"))
> ecoregions = st_transform(ecoregions, st_crs(landform))
> 
> # prepare cont data -------------------------------------------------------
> set.seed(222)
> random_ndvi = landcover
> random_ndvi$ndvi = sample(x = 1:10, size = length(random_ndvi[[1]]), replace = TRUE)
> random_ndvi$landcover2015.tif = NULL
> 
> # prepare sample landscapes -----------------------------------------------
> ext = st_bbox(c(xmin = -249797.344531127, xmax = -211162.693944285,
+                 ymin = -597280.143035389, ymax = -558645.492448547),
+                 crs = st_crs(landcover))
> 
> landcover_ext = landcover[ext]
> landform_ext = landform[ext]
> random_ndvi_ext = random_ndvi[ext]
> 
> # prepare my fun ----------------------------------------------------------
> my_fun = function(x) sum(!is.na(c(x[[1]])))
> 
> test_check("motif")
══ Failed tests ════════════════════════════════════════════════════════════════
── Error (test-lsp_add_clusters.R:47:1): (code run outside of `test_that()`) ───
Error in `UseMethod("lsp_add_examples")`: no applicable method for 'lsp_add_examples' applied to an object of class "c('grouped_df', 'tbl_df', 'tbl', 'data.frame')"
Backtrace:
    ▆
 1. └─motif::lsp_add_examples(x = landform_grid_sf_sel, y = landform) at test-lsp_add_clusters.R:47:0

[ FAIL 1 | WARN 0 | SKIP 0 | PASS 52 ]
Error: Test failures
Execution halted

1 error x | 0 warnings ✓ | 1 note x
````

**Comment by romainfrancois** [comment]
It seems to be because `slice_*()` no longer call `slice()` so `slice.sf` is not used when e.g. `slice_sample()` is used on an `sf` object: 

``` r
library(sf)
#> Linking to GEOS 3.9.0, GDAL 3.2.1, PROJ 7.2.1
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

nc <- st_read(system.file("shape/nc.shp", package="sf"))
#> Reading layer `nc' from data source 
#>   `/Users/romainfrancois/.R/library/4.0/sf/shape/nc.shp' using driver `ESRI Shapefile'
#> Simple feature collection with 100 features and 14 fields
#> Geometry type: MULTIPOLYGON
#> Dimension:     XY
#> Bounding box:  xmin: -84.32385 ymin: 33.88199 xmax: -75.45698 ymax: 36.58965
#> Geodetic CRS:  NAD27
nc$area_cl = cut(nc$AREA, c(0, .1, .12, .15, .25))

# ok because slice.sf()
nc %>% group_by(area_cl) %>% slice(1:2) %>% class()
#> [1] "sf"         "grouped_df" "tbl_df"     "tbl"        "data.frame"

# these non longer call slice() 
nc %>% group_by(area_cl) %>% slice_sample(n = 2) %>% class()
#> [1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
nc %>% group_by(area_cl) %>% slice_head(n = 2) %>% class()
#> [1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
nc %>% group_by(area_cl) %>% slice_tail(n = 2) %>% class()
#> [1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
nc %>% group_by(area_cl) %>% slice_min(PERIMETER, n = 2) %>% class()
#> [1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
nc %>% group_by(area_cl) %>% slice_max(PERIMETER, n = 2) %>% class()
#> [1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
```

<sup>Created on 2021-12-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by DavisVaughan** [comment]
Fixed in dplyr 1.0.8 by https://github.com/tidyverse/dplyr/pull/6126

``` r
library(sf)
#> Linking to GEOS 3.9.1, GDAL 3.4.0, PROJ 8.1.1; sf_use_s2() is TRUE
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

nc <- st_read(system.file("shape/nc.shp", package="sf"))
#> Reading layer `nc' from data source 
#>   `/Users/davis/Library/R/x86_64/4.1/library/sf/shape/nc.shp' 
#>   using driver `ESRI Shapefile'
#> Simple feature collection with 100 features and 14 fields
#> Geometry type: MULTIPOLYGON
#> Dimension:     XY
#> Bounding box:  xmin: -84.32385 ymin: 33.88199 xmax: -75.45698 ymax: 36.58965
#> Geodetic CRS:  NAD27
nc$area_cl = cut(nc$AREA, c(0, .1, .12, .15, .25))

# ok because slice.sf()
nc %>% group_by(area_cl) %>% slice(1:2) %>% class()
#> [1] "sf"         "grouped_df" "tbl_df"     "tbl"        "data.frame"

# these non longer call slice() 
nc %>% group_by(area_cl) %>% slice_sample(n = 2) %>% class()
#> [1] "sf"         "grouped_df" "tbl_df"     "tbl"        "data.frame"
nc %>% group_by(area_cl) %>% slice_head(n = 2) %>% class()
#> [1] "sf"         "grouped_df" "tbl_df"     "tbl"        "data.frame"
nc %>% group_by(area_cl) %>% slice_tail(n = 2) %>% class()
#> [1] "sf"         "grouped_df" "tbl_df"     "tbl"        "data.frame"
nc %>% group_by(area_cl) %>% slice_min(PERIMETER, n = 2) %>% class()
#> [1] "sf"         "grouped_df" "tbl_df"     "tbl"        "data.frame"
nc %>% group_by(area_cl) %>% slice_max(PERIMETER, n = 2) %>% class()
#> [1] "sf"         "grouped_df" "tbl_df"     "tbl"        "data.frame"
```

<sup>Created on 2022-06-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>
